### ISSUE TITLE
StringConcatenationToEncapsulation fixer

### ISSUE THREAD

**Issue by Soullivaneuh** [FriendsOfPHP/PHP-CS-Fixer#2334]
Regarding this [article](https://blog.blackfire.io/php-7-performance-improvements-encapsed-strings-optimization.html), it looks like using string encapsulation instead of concatenation provide better performance under PHP 7.

> The concatenation operator is not optimized. If you are using string concatenations, you end up doing the same weird things PHP 5 did. If you use encapsed strings, you will benefit from PHP 7 new string analysis algorithm, that performs the evaluated concatenation using what’s called a “Rope“.

I think having a fixer replacing concatenation by encapsulation when it cans would be a great addition.

What do you think?

**Comment by Slamdunk** [comment]
On PHP ^7.0:
```
$ time php -r '$y=str_repeat("g",1024);$z=uniqid();for($i=0;$i<100000000;++$i) $a="a ".$y." b c ".$z;'

real    0m13.252s
user    0m13.233s
sys     0m0.001s

$ time php -r '$y=str_repeat("g",1024);$z=uniqid();for($i=0;$i<100000000;++$i) $a="a $y b c $z";'

real    0m11.669s
user    0m11.651s
sys     0m0.002s
```
Yes, there is a gain, but I find string encapsulation a bad practice:

1. It's slower to read and you have to pay attention which quote is in use, single or double
1. It's harder to maintain because a variable and a string can't be concatenated, `$a . 'b' !== "$ab"`
1. IDE can't do proper highlighting

For similar reasons, in our business we also dropped concatenation in favour of `sprintf`.

**Comment by julienfalque** [comment]
I don't agree:

1. Whether encapsed strings are easier to read is a matter of being used to it IMO, just like not everyone could read PSR-2 compliant code easily before they started applying it
2. Not all cases can be converted by just removing the concatenation operator, but the fixer can just ignore those
3. I know at least PHPStorm can ;)

`sprintf` is useful but since I saw the linked blog post, I use encapsed strings whenever possible so I'm :+1: for such fixer.

**Comment by keradus** [comment]
:+1: for idea

**Comment by Soullivaneuh** [comment]
@Slamdunk Also note that encapsulation can (should?) be written like this:

```php
$a = 'test';
$b = "We are encapsulating {$a}";
```

And it's quite well managed by some IDE like PHPStorm:

![image](https://user-images.githubusercontent.com/1698357/32187355-17589248-bda5-11e7-80eb-fafa8b015b8d.png)

@juliendufresne @keradus BTW, if the fixer found something like this:

```php
$b = "We are encapsulating $a";
```

Should it convert it to:

```php
$b = "We are encapsulating {$a}";
```

Or leave it as is? Maybe this is the work for an another fixer.

**Comment by Slamdunk** [comment]
For encapsulated strings there already is a fixer proposed: https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3076

**Comment by OldStarchy** [comment]
Should I bump this issue or create a new one?

**Comment by Wirone** [comment]
@OldStarchy you should vote on existing issue with 👍 or provide PR 😉. Creating new issue or adding comment without actual value only brings noise.

**Comment by Wirone** [comment]
Personally I think it does not make much sense to invest time in this. Similarly to [sprintf → interpolation](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/3707#issuecomment-1601749275), there are several scenarios where it wouldn't be possible to refactor concatenation, I believe there would be more work related to determining if concatenation can be safely replaced with interpolation, than on actual refactoring. Benchmarks from cited post also are most probably outdated (it's been 7 years). Also, concatenation have one significant advantage - parts of the string can be divided into multiline code while keeping the string single line, and changing it to interpolation would end up with pretty long string (less readable). Personally I would close this issue since there wasn't much interest during these years.

**Comment by jlecordier** [comment]
Dang it, I guess I gotta write a custom rule then !

**Comment by Wirone** [comment]
@jlecordier thanks for reporting problems with the cookbook, I've created an issue for that.

In terms of a draft, I believe it's not possible. Please follow the existing fixers on how to manipulate the tokens, it's the best way to learn 😊.

**Comment by jlecordier** [comment]
Here's a draft, along with the tests. Feel free to say that it is shit (I already find it shitty myself 🤣)

```php
<?php

declare(strict_types=1);

/*
 * This file is part of PHP CS Fixer.
 *
 * (c) Fabien Potencier <fabien@symfony.com>
 *     Dariusz Rumiński <dariusz.ruminski@gmail.com>
 *
 * This source file is subject to the MIT license that is bundled
 * with this source code in the file LICENSE.
 */

namespace PhpCsFixer\Fixer\Operator;

use PhpCsFixer\AbstractFixer;
use PhpCsFixer\FixerDefinition\CodeSample;
use PhpCsFixer\FixerDefinition\FixerDefinition;
use PhpCsFixer\Tokenizer\CT;
use PhpCsFixer\Tokenizer\Token;
use PhpCsFixer\Tokenizer\Tokens;

/**
 * @author Jean LECORDIER <jean.lecordier@infoclimat.fr>
 */
final class StringConcatenationToInterpolationFixer extends AbstractFixer
{
    public function getDefinition(): FixerDefinition
    {
        // Return a definition of the fixer, it will be used in the documentation.
        return new FixerDefinition(
            'Transform a string concatenation by a string interpolation.', // Trailing dot is important. We thrive to use English grammar properly.
            [
                new CodeSample(
                    "<?php \$var = 'var'; \$str = 'string ' . \$var;\n",
                ),
            ]
        );
    }

    public function isCandidate(Tokens $tokens): bool
    {
        // Check whether the collection is a candidate for fixing.
        // Has to be ultra cheap to execute.
        return $tokens->isTokenKindFound('.');
    }

    protected function applyFix(\SplFileInfo $file, Tokens $tokens): void
    {
        // Add the fixing logic of the fixer here.
        $index = 0;
        while ($index < count($tokens)) {
            $token = $tokens[$index];
            $has_changed = false;
            if (is_variable($token)) {
                $has_changed = $this->handleVariableSomethingCase($tokens, $index);
            } elseif (is_simple_string($token)) {
                $has_changed = $this->handleStringSomethingCase($tokens, $index);
            } elseif (is_interpolation_quote($token)) {
                $has_changed = $this->handleInterpSomethingCase($tokens, $index);
            }
            if (!$has_changed) {
                $index++;
            }
        }
        $this->removeUselessConcatenation($file, $tokens);
    }

    private function removeUselessConcatenation(\SplFileInfo $file, Tokens $tokens): void
    {
        $concat_fixer = new NoUselessConcatOperatorFixer();
        $concat_fixer->configure(['juggle_simple_strings' => true]);
        $concat_fixer->applyFix($file, $tokens);
    }

    /**
     * Transforms `$var . ' string'` to `"{$var} string"`
     * or `$var . " {$interp}"` to `"{$var} {$interp}"`
     */
    private function handleVariableSomethingCase(Tokens $tokens, int $variable_index): bool
    {
        $variable_end_index = get_variable_end_index($tokens, $variable_index);

        $next_token_index = $tokens->getNextMeaningfulToken($variable_end_index);
        $next_token = $tokens[$next_token_index];
        if (!is_concatenation($next_token)) {
            return false;
        }
        $concat_index = $next_token_index;

        $next_next_token_index = $tokens->getNextMeaningfulToken($concat_index);
        $next_next_token = $tokens[$next_next_token_index];

        // Transforms `$var . ' string'` to `"{$var} string"`
        if (is_simple_string($next_next_token)) {
            return $this->handleVariableStringCase(
                $tokens,
                $variable_index,
                $variable_end_index,
                $next_next_token_index
            );
        }

        // Transforms `$var . " {$interp}"` to `"{$var} {$interp}"`
        if (is_interpolation_quote($next_next_token)) {
            return $this->handleVariableInterpCase(
                $tokens,
                $variable_index,
                $variable_end_index,
                $next_next_token_index
            );
        }

        return false;
    }

    /**
     * Transforms `$var . ' string'` to `"{$var} string"`
     */
    private function handleVariableStringCase(
        Tokens $tokens,
        int $variable_index,
        int $variable_end_index,
        int $right_string_index
    ): bool {
        $right_string_token = $tokens[$right_string_index];

        // The closing part must be done first, because the opening part will change the indexes
        // Replaces `$var . ' string'` by `$var} string"`
        $tokens->overrideRange(
            $variable_end_index + 1,
            $right_string_index,
            [
                get_close_curly_token(),
                get_raw_string_token($right_string_token),
                get_open_interpolation_token(),
            ]
        );

        // Replaces `$var` by `"{$var`
        $tokens->insertAt(
            $variable_index,
            [
                get_open_interpolation_token(),
                get_open_curly_token(),
            ]
        );

        return true;
    }

    /**
     * Transforms `$var . " {$interp}"` to `"{$var} {$interp}"`
     */
    private function handleVariableInterpCase(
        Tokens $tokens,
        int $variable_index,
        int $variable_end_index,
        int $open_quote_index
    ): bool {
        // The closing part must be done first, because the opening part will change the indexes
        // Replaces `$var . "` by `$var}`
        $tokens->overrideRange(
            $variable_end_index + 1,
            $open_quote_index,
            [get_close_curly_token()]
        );

        // Replaces `$var . "` by `"{$var`
        $tokens->insertAt(
            $variable_index,
            [
                get_open_interpolation_token(),
                get_open_curly_token(),
            ]
        );

        return true;
    }

    /**
     * Transforms `'string ' . $var` to `"string {$var}"`
     * or `'string ' . "{$interp}"` to `"string {$interp}"`
     */
    private function handleStringSomethingCase(Tokens $tokens, int $left_string_index): bool
    {
        $next_token_index = $tokens->getNextMeaningfulToken($left_string_index);
        $next_token = $tokens[$next_token_index];
        if (!is_concatenation($next_token)) {
            return false;
        }
        $concat_index = $next_token_index;

        $next_next_token_index = $tokens->getNextMeaningfulToken($concat_index);
        $next_next_token = $tokens[$next_next_token_index];

        // Transforms `'string ' . $var` to `"string {$var}"`
        if (is_variable($next_next_token)) {
            return $this->handleStringVariableCase(
                $tokens,
                $left_string_index,
                $next_next_token_index
            );
        }

        // Transforms `'string ' . "{$interp}"` to `"string {$interp}"`
        if (is_interpolation_quote($next_next_token)) {
            return $this->handleStringInterpCase(
                $tokens,
                $left_string_index,
                $next_next_token_index
            );
        }

        return false;
    }

    /**
     * Transforms `'string ' . $var` to `"string {$var}"`
     */
    private function handleStringVariableCase(
        Tokens $tokens,
        int $left_string_index,
        int $variable_index
    ): bool {
        $left_string_token = $tokens[$left_string_index];
        $variable_end_index = get_variable_end_index($tokens, $variable_index);

        // The closing part must be done first, because the opening part will change the indexes
        // Replaces `$var` by `$var}"`
        $tokens->insertAt(
            $variable_end_index + 1,
            [
                get_close_curly_token(),
                get_open_interpolation_token(),
            ]
        );

        // Replaces `'string ' . ` by `"string {`
        $tokens->overrideRange(
            $left_string_index,
            $variable_index - 1,
            [
                get_open_interpolation_token(),
                get_raw_string_token($left_string_token),
                get_open_curly_token(),
            ]
        );

        return true;
    }

    /**
     * Transforms `'string ' . "{$interp}"` to `"string {$interp}"`
     */
    private function handleStringInterpCase(
        Tokens $tokens,
        int $left_string_index,
        int $open_quote_index
    ): bool {
        $left_string_token = $tokens[$left_string_index];

        // Replaces `'string ' . "` by `"string `
        $tokens->overrideRange(
            $left_string_index,
            $open_quote_index,
            [
                get_open_interpolation_token(),
                get_raw_string_token($left_string_token),
            ]
        );

        return true;
    }

    /**
     * Transforms `"{$interp} " . $var` to `"{$interp} {$var}"`
     * or `"{$interp} " . 'string'` to `"{$interp} string"`
     */
    private function handleInterpSomethingCase(Tokens $tokens, int $open_quote_index): bool
    {
        $close_quote_token_index = get_closing_quote_index($tokens, $open_quote_index);
        if ($close_quote_token_index === null) {
            return false;
        }

        $next_token_index = $tokens->getNextMeaningfulToken($close_quote_token_index);
        $next_token = $tokens[$next_token_index];
        if (!is_concatenation($next_token)) {
            return false;
        }
        $concat_token_index = $next_token_index;

        $next_next_token_index = $tokens->getNextMeaningfulToken($concat_token_index);
        $next_next_token = $tokens[$next_next_token_index];

        // Transforms `"{$interp} " . $var` to `"{$interp} {$var}"`
        if (is_variable($next_next_token)) {
            return $this->handleInterpVariableCase(
                $tokens,
                $close_quote_token_index,
                $next_next_token_index
            );
        }

        // Transforms `"{$interp} " . 'string'` to `"{$interp} string"`
        if (is_simple_string($next_next_token)) {
            return $this->handleInterpStringCase(
                $tokens,
                $close_quote_token_index,
                $next_next_token_index
            );
        }

        return false;
    }

    /**
     * Transforms `"{$interp} " . $var` to `"{$interp} {$var}"`
     */
    private function handleInterpVariableCase(
        Tokens $tokens,
        int $close_quote_token_index,
        int $variable_index
    ): bool {
        $variable_end_index = get_variable_end_index($tokens, $variable_index);

        // The closing part must be done first, because the opening part will change the indexes
        // Replaces `$var` by `$var}"`
        $tokens->insertAt(
            $variable_end_index + 1,
            [
                get_close_curly_token(),
                get_open_interpolation_token(),
            ]
        );

        // Replaces `" . $var` by `{$var`
        $tokens->overrideRange(
            $close_quote_token_index,
            $variable_index - 1,
            [get_open_curly_token()]
        );

        return true;
    }

    /**
     * Transforms `"{$interp} " . 'string"` to `"{$interp} string"`
     */
    private function handleInterpStringCase(
        Tokens $tokens,
        int $close_quote_token_index,
        int $right_string_index
    ): bool {
        $right_string_token = $tokens[$right_string_index];

        // Replaces `" . 'string'` by `string"`
        $tokens->overrideRange(
            $close_quote_token_index,
            $right_string_index,
            [
                get_raw_string_token($right_string_token),
                get_open_interpolation_token()
            ]
        );

        return true;
    }
}

function is_concatenation(Token $token): bool
{
    return $token->getContent() === '.';
}

function is_variable(Token $token): bool
{
    return $token->isGivenKind(T_VARIABLE);
}

function is_simple_string(Token $token): bool
{
    return $token->isGivenKind(T_CONSTANT_ENCAPSED_STRING);
}

function is_interpolation_quote(Token $token): bool
{
    return $token->getContent() === '"';
}

function is_array_access_opening(Token $token): bool
{
    return $token->getContent() === '[';
}

/**
 * Transforms `'string'` to `string`
 */
function get_T_CONSTANT_ENCAPSED_STRING_content(Token $token): string
{
    return substr($token->getContent(), 1, -1);
}

/**
 * Transforms `'string'` to `string`
 */
function get_raw_string_token(Token $token): Token
{
    return new Token([
        T_ENCAPSED_AND_WHITESPACE,
        get_T_CONSTANT_ENCAPSED_STRING_content($token)
    ]);
}

function get_open_curly_token(): Token
{
    return new Token([T_CURLY_OPEN, '{']);
}

function get_close_curly_token(): Token
{
    return new Token([CT::T_CURLY_CLOSE, '}']);
}

function enclose_variable_token(Token $variable_token): array
{
    return [
        get_open_curly_token(),
        $variable_token,
        get_close_curly_token()
    ];
}

function get_interpolation_boundary_token(): Token
{
    return new Token('"');
}

function get_open_interpolation_token(): Token
{
    return get_interpolation_boundary_token();
}

function get_close_interpolation_token(): Token
{
    return get_interpolation_boundary_token();
}

function quote_interpolation(array $tokens): array
{
    return [
        get_open_interpolation_token(),
        ...$tokens,
        get_close_interpolation_token()
    ];
}

function get_closing_quote_index(Tokens $tokens, int $open_quote_index): ?int
{
    return $tokens->getNextTokenOfKind($open_quote_index, ['"']);
}

function get_closing_array_access_index(Tokens $tokens, int $array_access_opening_index): int
{
    return $tokens->getNextTokenOfKind($array_access_opening_index, [']']);
}

function get_variable_end_index(Tokens $tokens, int $variable_start_index): int
{
    $variable_end_index = $variable_start_index;
    $next_token_index = $tokens->getNextMeaningfulToken($variable_end_index);
    $next_token = $tokens[$next_token_index];
    if (!is_array_access_opening($next_token)) {
        return $variable_end_index;
    }
    $array_access_opening_index = $next_token_index;
    $variable_end_index = get_closing_array_access_index($tokens, $array_access_opening_index);
    return get_variable_end_index($tokens, $variable_end_index);
}
```

```php
<?php

declare(strict_types=1);

/*
 * This file is part of PHP CS Fixer.
 *
 * (c) Fabien Potencier <fabien@symfony.com>
 *     Dariusz Rumiński <dariusz.ruminski@gmail.com>
 *
 * This source file is subject to the MIT license that is bundled
 * with this source code in the file LICENSE.
 */

namespace PhpCsFixer\Tests\Fixer\Operator;

use PhpCsFixer\Tests\Test\AbstractFixerTestCase;

/**
 * @author Jean LECORDIER <jean.lecordier@infoclimat.fr>
 *
 * @internal
 *
 * @covers \PhpCsFixer\Fixer\Operator\StringConcatenationToInterpolationFixer
 */
final class StringConcatenationToInterpolationFixerTest extends AbstractFixerTestCase
{
    /**
     * @dataProvider provideFixCases
     */
    public function testFix(string $expected, ?string $input = null): void
    {
        $this->doTest($expected, $input);
    }

    public static function provideFixCases(): iterable
    {
        return [
            ['<?php $str = \'str\';'],
            ['<?php $str = "str";'],
            ['<?php $str = "left " . date("c");'],
            ['<?php $str = date("c") . " right";'],
            ['<?php $str = "left " . date("c") . " right";'],
            ['<?php $str = "left " . M_PI;'],
            ['<?php $str = M_PI . " right";'],
            ['<?php $str = "left " . M_PI . " right";'],
            ['<?php $str = "left " . __DIR__;'],
            ['<?php $str = __DIR__ . " right";'],
            ['<?php $str = "left " . __DIR__ . " right";'],
            ['<?php $var = "var"; $str = "left {$var}";'],
            ['<?php $var = "var"; $str = "{$var} right";'],
            ['<?php $var = "var"; $str = "left {$var} right";'],
            ['<?php $tab = ["tab"]; $str = "left {$tab[0]}";'],
            ['<?php $tab = ["tab"]; $str = "{$tab[0]} right";'],
            ['<?php $tab = ["tab"]; $str = "left {$tab[0]} right";'],
            [<<<'PHP'
                <?php $tab = ["key" => "value"]; $str = "left {$tab['key']}";
            PHP],
            [<<<'PHP'
                <?php $tab = ["key" => "value"]; $str = "{$tab['key']} right";
            PHP],
            [<<<'PHP'
                <?php $tab = ["key" => "value"]; $str = "left {$tab['key']} right";
            PHP],
            [
                '<?php $var = "var"; $str = "left {$var}";',
                '<?php $var = "var"; $str = "left " . $var;',
            ],
            [
                '<?php $var = "var"; $str = "{$var} right";',
                '<?php $var = "var"; $str = $var . " right";',
            ],
            [
                '<?php $var = "var"; $str = "left {$var} {$var}";',
                '<?php $var = "var"; $str = "left {$var} " . $var;',
            ],
            [
                '<?php $var = "var"; $str = "{$var} {$var} right";',
                '<?php $var = "var"; $str = $var . " {$var} right";',
            ],
            [
                '<?php $var = "var"; $str = "left {$var} right";',
                '<?php $var = "var"; $str = "left {$var}" . " right";',
            ],
            [
                '<?php $var = "var"; $str = "left {$var} right";',
                '<?php $var = "var"; $str = "left " . "{$var} right";',
            ],
            [
                '<?php $var = "var"; $str = "left {$var} right";',
                '<?php $var = "var"; $str = "left " . $var . " right";',
            ],
            [
                '<?php $tab = ["tab"]; $str = "left {$tab[0]}";',
                '<?php $tab = ["tab"]; $str = "left " . $tab[0];',
            ],
            [
                '<?php $tab = ["tab"]; $str = "{$tab[0]} right";',
                '<?php $tab = ["tab"]; $str = $tab[0] . " right";',
            ],
            [
                '<?php $tab = ["tab"]; $str = "left {$tab[0]} right";',
                '<?php $tab = ["tab"]; $str = "left " . $tab[0] . " right";',
            ],
            [
                '<?php $tab = [["tab"]]; $str = "left {$tab[0][0]}";',
                '<?php $tab = [["tab"]]; $str = "left " . $tab[0][0];',
            ],
            [
                '<?php $tab = [["tab"]]; $str = "{$tab[0][0]} right";',
                '<?php $tab = [["tab"]]; $str = $tab[0][0] . " right";',
            ],
            [
                '<?php $tab = [["tab"]]; $str = "left {$tab[0][0]} right";',
                '<?php $tab = [["tab"]]; $str = "left " . $tab[0][0] . " right";',
            ],
            [
                '<?php $tab = ["tab"]; $str = "{$tab[0]} {$tab[0]}";',
                '<?php $tab = ["tab"]; $str = "{$tab[0]} " . $tab[0];',
            ],
            [
                '<?php $tab = ["tab"]; $str = "{$tab[0]} {$tab[0]}";',
                '<?php $tab = ["tab"]; $str = $tab[0] . " {$tab[0]}";',
            ],
            [
                '<?php $tab = ["tab"]; $str = "{$tab[0]} {$tab[0]} {$tab[0]}";',
                '<?php $tab = ["tab"]; $str = "{$tab[0]} " . $tab[0] . " {$tab[0]}";',
            ],
            [
                '<?php $tab = [["tab"]]; $str = "{$tab[0][0]} {$tab[0][0]}";',
                '<?php $tab = [["tab"]]; $str = "{$tab[0][0]} " . $tab[0][0];',
            ],
            [
                '<?php $tab = [["tab"]]; $str = "{$tab[0][0]} {$tab[0][0]}";',
                '<?php $tab = [["tab"]]; $str = $tab[0][0] . " {$tab[0][0]}";',
            ],
            [
                '<?php $tab = [["tab"]]; $str = "{$tab[0][0]} {$tab[0][0]} {$tab[0][0]}";',
                '<?php $tab = [["tab"]]; $str = "{$tab[0][0]} " . $tab[0][0] . " {$tab[0][0]}";',
            ],
            [
                <<<'PHP'
                    <?php $tab = ["key" => "value"]; $str = "left {$tab['key']}";
                PHP,
                <<<'PHP'
                    <?php $tab = ["key" => "value"]; $str = "left " . $tab['key'];
                PHP,
            ],
            [
                <<<'PHP'
                    <?php $tab = ["key" => "value"]; $str = "{$tab['key']} right";
                PHP,
                <<<'PHP'
                    <?php $tab = ["key" => "value"]; $str = $tab['key'] . " right";
                PHP,
            ],
            [
                <<<'PHP'
                    <?php $tab = ["key" => "value"]; $str = "left {$tab['key']} right";
                PHP,
                <<<'PHP'
                    <?php $tab = ["key" => "value"]; $str = "left " . $tab['key'] . " right";
                PHP,
            ],
        ];
    }
}
```

**Comment by jlecordier** [comment]
I may add that such a rule exists in ESLint : https://eslint.org/docs/latest/rules/prefer-template

I understand the concerns stated above, but I think that the user should be able to make a choice, not the contributors of the fixers

I'll be glad to make a PR with the code presented above 😉

**Comment by Wirone** [comment]
@jlecordier please create a draft PR anyway, since it's much better to review than a comment in the closed issue 🙂. We will have then CI feedback too, because right now I don't even know if it works.
